staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"fmt"
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/checker/decls"
+	expr "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	celmodel "k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model"
+)
+
+// ScopedTypeName is the placeholder type name used for the type of ScopedVarName if it is an object type.
+const ScopedTypeName = "apiextensions.k8s.io.v1alpha1.ValidationExpressionSelf"
+
+// ScopedVarName is the variable name assigned to the locally scoped data element of a CEL valid.
+const ScopedVarName = "self"
+
+// Compile compiles all the CEL validation rules in the CelRules and returns a slice containing a compiled program for each provided CelRule, or an array of errors.
+func Compile(s *schema.Structural) ([]cel.Program, []error) {
+	var allErrors []error
+	if len(s.Extensions.XValidations) == 0 {
+		return nil, nil
+	}
+	celRules := s.Extensions.XValidations
+
+	var propDecls []*expr.Decl
+	var root *celmodel.DeclType
+	var ok bool
+	env, _ := cel.NewEnv()
+	reg := celmodel.NewRegistry(env)
+	rt, err := celmodel.NewRuleTypes(ScopedTypeName, s, reg)
+	if err != nil {
+		allErrors = append(allErrors, err)
+		return nil, allErrors
+	}
+	opts, err := rt.EnvOptions(env.TypeProvider())
+	if err != nil {
+		allErrors = append(allErrors, err)
+		return nil, allErrors
+	}
+	root, ok = rt.FindDeclType(ScopedTypeName)
+	if !ok {
+		root = celmodel.SchemaDeclType(s).MaybeAssignTypeName(ScopedTypeName)
+	}
+	// if the type is object, will traverse each field in the object tree and declare
+	if root.IsObject() {
+		for k, f := range root.Fields {
+			// TODO: There is a much larger set of identifiers that collide if unnested
```

Comment:
shouldn't we decide for one option as a pre-req of persisting any data into etcd ?

Reply:
My taste:
- 3 is ugly
- 2 is a little better, but requires that the reserved key words never change, and we must document which these are.
- 1: which escape mechanism would you use?

Reply:
I thought the KEP already proposed to use `_`. Although, I think using `self` is more clear.

Reply:
The escaping implementation would be added in https://github.com/kubernetes/kubernetes/pull/106051

Reply:
I've lost access to the list of options that kicked off this thread, so I don't know what exact options are being compared, but for escaping, my current thinking is:

- We have to deconflict with "RESERVED" CEL lexer keywords somehow. My proposal is that we prefix with _.  This will be most noticeable with 'namespace', which is both a reserved keyword in CEL and a core concept in kubernetes.  To ensure prefixed identifer doesn't collide with other field names, we prefix anything prefixed with _ with N+1 _s.  This rule applies to all field names used anywhere (since the lexer is aware of grammatical context).

That's the easy part, I think. The harder part is what we do about the property names that we binds as variables into the root expression scope. The options for this are:

1. We stop binding properties in the root scope and instead require all properties be accessed via 'self'.
2. We bind property names IF they don't conflict with any other identifiers (incl type names, function names and macro names). If they conflict, we just don't make them available in the root scope.  (This is what we have in the implementation now)
3. We bind property names and leave it up to the parser to determine if they are accessible. 
4. We bind property names, but if they conflict with any other identifiers, we escape them.

I'm currently favoring #1 right now, but it would require a KEP update. My second choice would be #3, but I want to do some testing to see how the parser would behave.

Reply:
I did some testing of CEL and found:

- CEL can differentiate between function/macro names (`has`, `size`, ...) and identifiers.
- CEL cannot differentiate between type identifiers (`int`, `string`, ...)

So we only need to deal with escaping for the lexers reserved keywords and the relatively small (and fixed) set of type identifiers. I've updated the escaper one my PR.

Reply:
> I'm currently favoring #1 right now, but it would require a KEP update.

Are you proposing that we do that forever, or just for a few releases? If just for a few releases, take in consideration that loosening validation in the API is a convoluted process.

IIUC, option 4 is what the KEP initially proposed, right? What is the problem with it now?